src/core/service.c

```diff
@@ -3028,6 +3025,20 @@ static void service_live_mount_finish(Service *s, ServiceResult f, const char *e
         s->mount_request = sd_bus_message_unref(s->mount_request);
 }
 
+static void service_reload_finish(Service *s, ServiceResult f) {
+        assert(s);
+
+        s->reload_result = f;
+        s->reload_begin_usec = USEC_INFINITY;
+
+        /* If notify state is still in dangling NOTIFY_RELOADING, reset it so service_enter_running()
+         * won't get confused (see #37515) */
+        if (s->notify_state == NOTIFY_RELOADING)
+                s->notify_state = _NOTIFY_STATE_INVALID;
```

Comment:
i wonder if this is racy? i.e. let's say an ExecReload= cmd fails with an unclean exit code, and we now end up here, but in the time the cmd failed and we come here the user already tried to reload the service another way, so we get another RELOADING=1 here. We'll now ignore htat, and miss the state change.

we could probably make this airtight, by only resetting things here if the timestamp of the msg predate the actual state change to reloading state?

or am i overthinking this?

Reply:
My conclusion so far is that ExecReload= is inherently racy. We simply don't know if the RELOADING=1 the main process sent us is due to ExecReload= doing its thing or from extrinsic event. It could be sent before executor handoff - sure we have handoff timestamps now but that's just part of the problem.

To main goal of this patchset is to make sure ExecReload= and ReloadSignal= work together consistently. Previously some states in between are ill-formed, now ExecReload= can be used for e.g. pre-reload businesses (config check?) as well as signaling the actual reload. If it's the former, the notify-reload protocol fires which is not racy; for the latter it's all about creativity ;)